numpy/random/_generator.pyx

```diff
@@ -4150,17 +4156,64 @@ cdef class Generator:
 
         i = 0
         totsize = np.PyArray_SIZE(val_arr)
-        with self.lock, nogil:
-            while i < totsize:
-                acc = 0.0
-                for j in range(k):
-                    val_data[i+j] = random_standard_gamma(&self._bitgen,
-                                                              alpha_data[j])
-                    acc = acc + val_data[i + j]
-                invacc = 1/acc
-                for j in range(k):
-                    val_data[i + j] = val_data[i + j] * invacc
-                i = i + k
+
+        # Small alpha case: Use stick-breaking approach with beta
+        # random variates (RVs).
+        #
+        # This approach prevents NaNs resulting from 0/0 that may
+        # occur when using the standard approach of generating
+        # dirichlet RVs via normalisation of a vector of gamma RVs
+        #
+        # In the standard approach each alpha value is used as the
+        # shape parameter of a gamma distribution. If all the alpha
+        # values are sufficiently small, there is a high probability
+        # that all the gamma variates will be 0. When that happens,
+        # the normalization process ends up computing 0/0, giving nan.
+        #
+        # The stick-breaking algorithm does not use divisions, so that
+        # a situation in which 0/0 has to be computed cannot occur.
+        #
+        # The algorithm is slower than the standard approach as
+        # generation of beta RVs is slower than generation of gamma
+        # RVs.
+        #
+        # The switch from the standard to the stick-breaking algorithm
+        # is made whenever the maximum of all alpha values is smaller
+        # than a threshold `t`. The threshold is chosen such that for
+        # `alpha.max() >= t` the probability that all gamma variates
+        # will be zero is neglegible.
+        #
+        # Formally for a given threshold `t` this probability can be
+        # bounded by `gammainc(t, d)` where `gammainc` is the
+        # regularized incomplete gamma function and `d` is the
+        # smallest positive floating point number that can be
+        # represented with a given precision.
+        if alpha_arr.max() < 0.1:
+            with self.lock, nogil:
+                while i < totsize:
+                    acc = 1.
+                    for j in range(k - 1):
+                        v = random_beta(&self._bitgen, alpha_data[j],
+                                        alpha_csum_data[j + 1])
+                        val_data[i + j] = acc * v
+                        acc *= (1. - v)
+                    val_data[i + j + 1] = acc
```

Comment:
This warning in the travis CI log indicates a real issue:

```
  numpy/random/_generator.c:19904:12: warning: ‘__pyx_v_j’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```

It corresponds to the Cython line

                    val_data[i + j + 1] = acc

that occurs *after* for the `j` loop.  The variable `j` is presumed to have been set to the final value from the `for`-loop.  But if `k` is 1, then `range(k - 1)` is empty, the `for`-loop does not execute, and `j` is not set.  The simple fix is to replace that line with

                    val_data[i + k - 1] = acc

Reply:
Thanks for the explanation. This is fixed now.